test/fgpu/test_engine.py

```diff
@@ -933,6 +937,60 @@ async def test_missing_heaps(
+        spectra_per_output_chunk = engine_server.chunk_jones // output.channels
+        heaps_per_output_chunk = spectra_per_output_chunk // spectra_per_heap
+
+        # Need to reshape timestamp list so that we can work in batches of
+        # chunk size. The last chunk isn't finished so we need to add a
+        # bit on the end.  Need to test this after the prometheus counters
+        # becaus this affects np arrays which those tests use to count things.
+        if (remainder := len(unix_timestamps) % spectra_per_output_chunk) != 0:
+            unix_timestamps.extend([0 for _ in range(spectra_per_output_chunk - remainder)])
+        chunk_timestamps = np.reshape(unix_timestamps, (-1, spectra_per_output_chunk))
+        chunk_time_step = np.average(chunk_timestamps[1]) - np.average(chunk_timestamps[0])
+
+        # Same with dst_present
+        if (remainder := len(dst_present) % heaps_per_output_chunk) != 0:
+            dst_present = np.append(dst_present, [False for _ in range(heaps_per_output_chunk - remainder)])
+        chunk_present = np.reshape(dst_present, (-1, heaps_per_output_chunk))
+
+        # Sensor is not present in the narrowband mode.
+        if output.decimation == 1:
+            # We can't just for-loop because we may not have an update every chunk.
+            sensor_update_iter = iter(sensor_update_dict["input1.dig-rms-dbfs"])  # Both sensors should be the same.
+            sensor_update = next(sensor_update_iter)
+
+            for chunk_time, present in zip(chunk_timestamps, chunk_present):
+                # Sensor timestamp should be close to the end of the chunk if # it exists.
+                diff = abs(sensor_update.timestamp - np.max(chunk_time))
+                if diff / chunk_time_step > 1:  # Somewhat arbitrary, if a chunk is skipped this will be large.
+                    # The next sensor update is some time in the future, so we
+                    # don't have anything to compare it to for a while.
+                    assert not np.any(present)
+                else:
+                    if np.any(present) or len(present) == 1:
+                        # There were heaps in this chunk, or the edge case where a chunk is one heap.
+                        if np.all(present):
+                            assert (
+                                sensor_update.status == aiokatcp.Sensor.Status.NOMINAL
+                                or sensor_update.status == aiokatcp.Sensor.Status.WARN
+                            )
+                        else:
+                            assert sensor_update.status == aiokatcp.Sensor.Status.FAILURE
+
+                        try:
+                            sensor_update = next(sensor_update_iter)
+                        except StopIteration:
+                            # We are finished with sensor updates. The last
+                            # chunk doesn't finish so we don't expect anymore.
+                            break
```

Comment:
Simply stopping the test here means that an implementation that gets the first update correct but never updates again is going to pass.

The way I see it, there are 3 cases for a chunk:
- np.all(present): an update MUST be present with NOMINAL or WARN status
- if not, but np.any(present): an update MUST be present with FAILURE status
- not np.any(present): a single update MAY be present, in which case it must have FAILURE status.

I don't think you're verifying all those conditions.